plugins/modules/vmware_guest.py

```diff
@@ -1667,38 +1683,42 @@ def configure_hardware_params(self, vm_obj):
                                               " not supported. Please specify version greater"
                                               " than the current version." % (version_digit,
                                                                               temp_version))
-                    new_version = "vmx-%02d" % temp_version
-                    try:
-                        task = vm_obj.UpgradeVM_Task(new_version)
-                        self.wait_for_task(task)
-                        if task.info.state == 'error':
-                            return {'changed': self.change_applied, 'failed': True, 'msg': task.info.error.msg, 'op': 'upgrade'}
-                    except vim.fault.AlreadyUpgraded:
-                        # Don't fail if VM is already upgraded.
-                        pass
+
+                    # Only perform the upgrade if not in check mode.
+                    # self.change_detected was already set above in the self.configspec.version != vm_obj.config.version check above
+                    if not self.module.check_mode:
+                        new_version = "vmx-%02d" % temp_version
+                        try:
+                            task = vm_obj.UpgradeVM_Task(new_version)
+                            self.wait_for_task(task)
+                            if task.info.state == 'error':
+                                return {'changed': self.change_applied, 'failed': True, 'msg': task.info.error.msg, 'op': 'upgrade'}
+                        except vim.fault.AlreadyUpgraded:
+                            # Don't fail if VM is already upgraded.
+                            pass
 
         secure_boot = self.params['hardware']['secure_boot']
         if secure_boot is not None:
-            self.configspec.bootOptions = vim.vm.BootOptions()
-            self.configspec.bootOptions.efiSecureBootEnabled = True
-            if vm_obj is None or self.configspec.bootOptions.efiSecureBootEnabled != vm_obj.config.bootOptions.efiSecureBootEnabled:
+            if vm_obj is None or secure_boot != vm_obj.config.bootOptions.efiSecureBootEnabled:
                 self.change_detected = True
+                self.configspec.bootOptions = vim.vm.BootOptions()
+                self.configspec.bootOptions.efiSecureBootEnabled = secure_boot
```

Comment:
note that I changed this from always setting to True to setting to True/False based on the value. Was it intended that it was always set to True before, or was that a bug?

Reply:
Yes, I've seen this. I should say this was a bug. Doesn't make sense to enable secure boot when someone defines it to be disabled ;-)